File: leetcode/python/21. 合并两个有序链表.py
#!/bin/bash python
"""
21. 合并两个有序链表
有两种思路：递归法，迭代法
可以原地修改链接，或者new新的链表
以下是原地修改的方式（破坏了原始的链表）
"""

class ListNode(object):
	def __init__(self, val=0, next=None):
		self.val = val
		self.next = next


class Solution(object):
	def mergeTwoLists(self, p, q):
		"""
		递归法
		"""
		if not p:
			return q
		if not q:
			return p
		if p.val < q.val:
			p.next = self.mergeTwoLists(p.next, q)
			return p
		else:
			q.next = self.mergeTwoLists(p, q.next)
			return q
	
	def mergeTwoLists2(self, p, q):
		"""
		迭代法
		"""
		if not p:
			return q
		if not q:
			return p
		newHead = ListNode()
		i, j, head = p, q, newHead
		while i and j:
			if i.val < j.val:
				head.next = i
				i = i.next
			else:
				head.next = j
				j = j.next
			head = head.next
		if i:
			head.next = i
		if j:
			head.next = j
		return newHead.next

    def mergeTwoLists(self, p, q):
        """
        :type p: ListNode
        :type q: ListNode
        :rtype: ListNode
        迭代的方式，经典
        """
        curr = newHead = ListNode(0)
        while p and q:
            if p.val < q.val:
                curr.next = p
                p = p.next
            else:
                curr.next = q
                q = q.next
            curr = curr.next
        curr.next = p if p else q
        return newHead.next

File: leetcode/cpp/21. 合并两个有序链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

#include <ListNode.h>

class Solution {
public:
    // 循环法
    ListNode* mergeTwoLists(ListNode* p, ListNode* q) {
        if (p == nullptr) {
            return q;
        }
        if (q == nullptr) {
            return p;
        }

        ListNode* head = new ListNode(0);
        ListNode* curr = head;
        while (p && q) {
            if (p->val < q->val) {
                curr->next = new ListNode(p->val);
                p = p->next;
            } else {
                curr->next = new ListNode(q->val);
                q = q->next;
            }
            curr = curr->next;
        }
        curr->next = p ? p : q;

        return head->next;
    }

	// 递归法
	ListNode* mergeTwoLists2(ListNode* p, ListNode* q) {
		if (p == nullptr) {
			return q;
		}
		if (q == nullptr) {
			return p;
		}
		if (p->val < q->val) {
			p->next = mergeTwoLists2(p->next, q);
			return p;
		} else {
			q->next = mergeTwoLists2(p, q->next);
			return q;
		}
	}

    ListNode* mergeTwoLists(ListNode* p, ListNode* q) {
        ListNode *newHead = new ListNode(), *curr = newHead;
        while (p && q) {
            if (p->val < q->val) {
                curr->next = p;
                p = p->next;
            } else {
                curr->next = q;
                q = q->next;
            }
            curr = curr->next;
        }
        curr->next = (p ? p : q);
        return newHead->next;
    }
};